src/box/index.cc

```diff
@@ -806,6 +806,20 @@ index_delete(struct index *index)
 	index_def_delete(def);
 }
 
+int
+index_replace_with_txm(struct index *index, struct tuple *old_tuple,
+		       struct tuple *new_tuple, enum dup_replace_mode mode,
+		       struct tuple **result, struct tuple **successor)
+{
+	memtx_tx_story_gc();
+	if (new_tuple == NULL)
+		return memtx_tx_history_add_delete(index, old_tuple, result);
+	return index->vtab->replace(index, NULL, new_tuple,
+				    DUP_REPLACE_OR_INSERT, result, successor) ||
+		memtx_tx_history_add_insert(index, old_tuple, new_tuple, mode,
```

Comment:
Nit: Hard to read. Better write explicitly:

```C
if (index->vtab->replace(...) != 0)
        return memtx_tx_history_add_insert(...);
return 0;
```

Reply:
> Hard to read.

I believe it's rather unusual than hard. I know I could go for the conventional option you suggest, but I like the style I suggest, because it looks like a pipeline notation (i.e., perform a sequence of actions propagating the result further and aborting the pipeline if any of the actions happen to fail).

Reply:
Nevermind, I decided to pull out the index state rollback in case of transaction manager notification failure, so I now use @locker's approach.